### DISCUSSION TITLE
Apply fields of the mapper to the target

### DISCUSSION THREAD

**Discussion by gr0l** [mapstruct/mapstruct#2924] [Ideas]
Hello community,

I would like to implement an AbstractMapper which has multiple source informations wich should be mapped to a complex object instance.  

My Idea is that it would be great if the name `this` or any other defined name can be used to directly address fields of the mapper itself. This would reduce the amount of self written code and parameters for the mapper.

``` java
public abstract class MyMapper {

  @Inject
  @Named("MySpecialName")
  String name;
  
  @Inject
  ComplexType someOtherInformation;
  
  
  @Mapping(target = "generalName", source = "this.name")
  @Mapping(target = "id", source = "this.someOtherInformation.id")
  @Mapping(target = "messageText", source = "text")
  abstract  Result map(Message message);
}
```

Currently I write a mapping function for each type or add multiple parameters to the mapper. And then I have to write a an method which combines the different mapping methods.

**Comment by Zegveld** [comment]
You can currently use expressions for these cases. 

for example:
* `expression = "java(name)"` instead of `source = "this.name"`
* `expression = "java(someOtherInformation.id)"` instead of `source = "this.someOtherInformation.id"`.
Or if the ID is accessable with a getter `expression = "java(someOtherInformation.getId())"`

I do like the idea, but I do not know what the impact will be of implementing this. The chances of someone having a field called `this` is very low but not completely absent as it is possible with `getThis()` and `setThis(...)` methods and a private field named `thiz` or something.

**Comment by filiphr** [comment]
I am not sure that I understand the use case of using injected fields of a mapper to perform a mapping. As @Zegveld suggested you can always use `expression` to use those fields.

In addition to that you can also use a wrapper object or multi source mapping. e.g.

```java
public abstract class MyMapper {

  @Inject
  @Named("MySpecialName")
  String name;
  
  @Inject
  ComplexType someOtherInformation;
  
  
  @Mapping(target = "generalName", source = "this.name")
  @Mapping(target = "id", source = "this.someOtherInformation.id")
  @Mapping(target = "messageText", source = "text")
  Result map(Message message) {
      return message != null ? map(message, name, someOtherInformation) : null;
  }

  @Mapping(target = "generalName", source = "name")
  @Mapping(target = "id", source = "someOtherInformation.id")
  @Mapping(target = "messageText", source = "message.text")
  abstract  Result map(Message message, String name, ComplexType someOtherInformation);
}
```
